jdk/src/share/classes/java/util/Formatter.java

```diff
@@ -4380,9 +4380,10 @@ private char getZero(Locale l) {
 
             // apply zero padding
             len = sb.length();
-            if (width != -1 && f.contains(Flags.ZERO_PAD))
-                for (int k = 0; k < width - len; k++)
-                    sb.insert(begin, zero);
+            if (width > len && f.contains(Flags.ZERO_PAD)) {
+                String zeros = new String(new char[width - len]).replace("\0", "0");
```

Comment:
In the original patch, we use the locale-dependent `zero` variable here. I don't think we can just use `0`, although I struggle to imagine the case where that would matter. Plus, given we are changing char->char, we can use `String.replace(char,char)` instead of `String.replace(CharSequence,CharSequence)`.

I think we can hit these two birds with one stone by doing:

```
 char[] zeros = new char[width - len];
 Arrays.fill(zeros, zero);
 sb.insert(begin, new String(zeros));
```

It is still awkward because `String` constructor copies the array. For extreme use cases, there is a package-private `String(char[] array, boolean share)` constructor that might be accessible through `SharedSecrets.getJavaLangAccess().newSharedString(...)` (needs a new method). But that might be moot point since `AbstractStringBuilder.insert` already does a lot of copies.